antarest/study/service.py

```diff
@@ -2731,26 +2709,48 @@ def delete_file_or_folder(self, study_id: str, path: str, current_user: JWTUser)
             current_user: User that called the endpoint
 
         Raises:
-            FileDeletionNotAllowed: if the path does not comply with the above rules
+            ResourceDeletionNotAllowed: if the path does not comply with the above rules
         """
-        study = self.get_study(study_id)
-        assert_permission(current_user, study, StudyPermissionType.WRITE)
+        args = {"path": path}
+        self._alter_user_folder(study_id, args, current_user, RemoveUserResource, ResourceDeletionNotAllowed)
 
-        url = [item for item in path.split("/") if item]
-        if len(url) < 2 or url[0] != "user":
-            raise FileDeletionNotAllowed(f"the targeted data isn't inside the 'User' folder: {path}")
+    def create_folder(self, study_id: str, path: str, current_user: JWTUser) -> None:
+        """
+        Creates a folder inside the study.
+        The data must be located inside the 'User' folder.
+        Also, it can not be inside the 'expansion' folder.
 
-        study_tree = self.storage_service.raw_study_service.get_raw(study, True).tree
-        user_node = t.cast(User, study_tree.get_node(["user"]))
-        if url[1] in [file.filename for file in user_node.registered_files]:
-            raise FileDeletionNotAllowed(f"you are not allowed to delete this resource : {path}")
+        Args:
+            study_id: UUID of the concerned study
+            path: Path corresponding to the resource to be deleted
+            current_user: User that called the endpoint
+
+        Raises:
+            FolderCreationNotAllowed: if the path does not comply with the above rules
+        """
+        args = {"path": path, "file": False}
+        self._alter_user_folder(study_id, args, current_user, CreateUserResource, FolderCreationNotAllowed)
+
+    def _alter_user_folder(
+        self,
+        study_id: str,
+        args: t.Dict[str, t.Any],
```

Comment:
We should use stronger typing here, the signature is not clear enough.
For example, we could have a small `CreateUserResourceData` class that contains `path` and `file`. This class could be used as the input to `CreateUserResource` creation and inside it.
This si actually a scheme that we could use for all command types, in order to use pydantic models for their associated data.

Reply:
I introduced such classes I'll let you take a look

Reply:
Nice but it would be better to use composition instead of inheritance here:
```python
class CreateUserResource(ICommand, CreateUserResourceData):
```